core/src/storage/device.py

```diff
@@ -151,6 +153,85 @@ def get_mnemonic_secret() -> bytes | None:
     return common.get(_NAMESPACE, _MNEMONIC_SECRET)
 
 
+def store_mnemonic_secret(
+    secret: bytes,
+    needs_backup: bool = False,
+    no_backup: bool = False,
+    allow_derivation_fail: bool = False,
+) -> None:
+    set_version(common.STORAGE_VERSION_CURRENT)
+    common.set(_NAMESPACE, _MNEMONIC_SECRET, secret)
+    common.set_true_or_delete(_NAMESPACE, _NO_BACKUP, no_backup)
+    common.set_bool(_NAMESPACE, INITIALIZED, True, public=True)
+    if not no_backup:
+        common.set_true_or_delete(_NAMESPACE, _NEEDS_BACKUP, needs_backup)
+
+    if not utils.BITCOIN_ONLY:
+        _store_binary_mnemonic(secret, allow_derivation_fail)
+
+
+if not utils.BITCOIN_ONLY:
+
+    def get_binary_mnemonic() -> bytes | None:
+        """
+        Get the binary representation of mnemonic (including checksum).
+        """
+        return common.get(_NAMESPACE, _BINARY_MNEMONIC)
+
+    def _store_binary_mnemonic(secret: bytes, allow_derivation_fail: bool) -> None:
+        """
+        Store the bianry representation of mnemonic (including checksum) for Cardano
+        Icarus derivation. Works only for BIP-39.
+
+        If `allow_derivation_fail` is True, exception during derivation is ignored.
+        """
+        from trezorcrypto import bip39
+
+        from trezor.enums import BackupType
+
+        if get_backup_type() == BackupType.Bip39:
+            try:
+                binary_mnemonic = bip39.mnemonic_to_bits(secret.decode())
+            except ValueError:
+                if not allow_derivation_fail:
+                    raise
+                else:
+                    return
+
+            common.set(
+                _NAMESPACE,
+                _BINARY_MNEMONIC,
+                binary_mnemonic,
+            )
+
+    def update_binary_mnemonic() -> bytes | None:
+        """
+        Sets the binary representation of mnemonic (including checksum) in storage.
+
+        Works only for BIP-39.
+
+        Returns:
+            `bytes | None`: The derived binary mnemonic.
+        Raises:
+            `RuntimeError`: If the stored mnemonic cannot be decoded as UTF-8, which
+            indicates the backup type is SLIP-39.
+        """
+        from trezorcrypto import bip39
+
+        secret = get_mnemonic_secret()
+        if secret is None:
+            return None
+
+        try:
+            secret_str = secret.decode()
+            binary_mnemonic = bip39.mnemonic_to_bits(secret_str)
+        except UnicodeError:
+            raise RuntimeError("Mnemonic to bits derivation called for SLIP39.")
+
+        common.set(_NAMESPACE, _BINARY_MNEMONIC, binary_mnemonic)
+        return binary_mnemonic
```

Comment:
This function does not need to return `binary_mnemonic` any more. Actually, it seems like the function could be completely replaced with `_store_binary_mnemonic()`. The `UnicodeError` handling shouldn't be necessary.